File: 0_1.baekjoon/brute-force/java/src/BOJ14391.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class BOJ14391 {

    static int N, M;
    static int[][] board;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());

        board = new int[N][M];

        for (int i = 0; i < N; i++) {
            String line = br.readLine();

            for (int j = 0; j < M; j++) {
                board[i][j] = line.charAt(j) - '0';
            }
        }

        System.out.println(dfs(0, 0));
    }

    private static int dfs(int idx, int check) {
        if (idx == (N * M)) {
            return getScore(check);
        }

        int result = dfs(idx + 1, check | (1 << idx)); // 가로
        result = Math.max(result, dfs(idx + 1, check)); // 세로

        return result;
    }

    private static int getScore(int visited) {
        int total = 0;

        for (int i = 0; i < N; i++) {
            int horizon = 0;
            for (int j = 0; j < M; j++) {
                if ((visited & (1 << (i * M + j))) != 0) {
                    horizon *= 10;
                    horizon += board[i][j];
                } else {
                    total += horizon;
                    horizon = 0;
                }
            }
            total += horizon;
        }

        for (int j = 0; j < M; j++) {
            int vertical = 0;
            for (int i = 0; i < N; i++) {
                if ((visited & (1 << (i * M + j))) == 0) {
                    vertical *= 10;
                    vertical += board[i][j];
                } else {
                    total += vertical;
                    vertical = 0;
                }
            }
            total += vertical;
        }

        return total;
    }
}

File: 0_1.baekjoon/brute-force/kotlin/src/BOJ14391.kt
import kotlin.math.max

private var N = 0
private var M = 0

private lateinit var board: Array<IntArray>

fun main() = with(System.`in`.bufferedReader()) {
    val (n, m) = readLine().split(" ").map { it.toInt() }

    N = n
    M = m

    board = Array(n) {
        readLine()!!.toCharArray().map { Character.getNumericValue(it) }.toIntArray()
    }

    print(dfs(0, 0))
}

private fun dfs(idx: Int, check: Int): Int {
    if (idx == N * M) {
        return getScore(check)
    }

    var result = dfs(idx + 1, check or (1 shl idx))
    result = max(result, dfs(idx + 1, check))

    return result
}

private fun getScore(check: Int): Int {
    var total = 0

    for (i in 0 until N) {
        var horizon = 0
        for (j in 0 until M) {
            if ((check and (1 shl i * M + j)) != 0) {
                horizon *= 10
                horizon += board[i][j]
            } else {
                total += horizon
                horizon = 0
            }
        }

        total += horizon
    }


    for (j in 0 until M) {
        var vertical = 0
        for (i in 0 until N) {
            if ((check and (1 shl i * M + j)) == 0) {
                vertical *= 10
                vertical += board[i][j]
            } else {
                total += vertical
                vertical = 0
            }
        }

        total += vertical
    }

    return total
}
